# Wrangling Data: WeRateDogs

In this project we are going to gather, asses and clean data from the WeRateDogs Twitter account. Then we will analyze this data and present some insights and visualizations.

In [1]:
# import libraries
import pandas as pd
import requests
import os
import tweepy
import json
%matplotlib inline

## Gathering data

1) Read the file provided by Udacity that contains data from the tweets of WeRateDogs

In [2]:
data = pd.read_csv('twitter-archive-enhanced.csv')

2) Download programmatically the image file and read it

In [3]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response=requests.get(url)
with open('./image-predictions.tsv',mode='wb') as file:
    file.write(response.content)

In [4]:
image_pred = pd.read_csv('image-predictions.tsv',sep='\t')

3) Query the Twitter API for each tweet's JSON data. Store in a file. read the file into a DataFrame

In [5]:
consumer_key = '...'
consumer_secret = '...'
access_token = '...'
access_secret = '...'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [6]:
# checking if works fine
tweet = api.get_status(892420643555336193)
print(tweet.text)

This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU


In [8]:
# works fine, lets gather the tweets ids from the data DataFrame
ids= data.tweet_id.values
print(len(ids))
fails = {}
check = 0
with open('./tweet_json.txt', 'w') as file:
    for tweet_id in ids:
        check += 1
        #print(check)
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            json.dump(tweet._json, file)
            file.write('\n')
        except tweepy.TweepError as error:
            fails[tweet_id] = error
            pass
print(fails)

2356
Rate limit reached. Sleeping for: 721
Rate limit reached. Sleeping for: 711
{888202515573088257: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 873697596434513921: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 872668790621863937: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 872261713294495745: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 869988702071779329: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 866816280283807744: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 861769973181624320: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 856602993587888130: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 851953902622658560: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 845459076796616705: TweepError([{'code': 144, 'message': 'No status foun

In [11]:
extra = pd.read_json('tweet_json.txt', lines=True)

We already have all the information in three DataFrames: data, image_pred and extra. 

## Assesing data

In [12]:
data.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [14]:
data.dtypes

tweet_id                        int64
in_reply_to_status_id         float64
in_reply_to_user_id           float64
timestamp                      object
source                         object
text                           object
retweeted_status_id           float64
retweeted_status_user_id      float64
retweeted_status_timestamp     object
expanded_urls                  object
rating_numerator                int64
rating_denominator              int64
name                           object
doggo                          object
floofer                        object
pupper                         object
puppo                          object
dtype: object

In [15]:
data.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [16]:
#all the denominators should be 10, lets take a look
data[data.rating_denominator != 10][['text','rating_denominator']]

text  rating_denominator
313   @jonnysun @Lin_Manuel ok jomny I know you're e...                   0
342            @docmisterio account started on 11/15/15                  15
433   The floofs have been released I repeat the flo...                  70
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                   7
784   RT @dog_rates: After so many requests, this is...                  11
902   Why does this never happen at my front door......                 150
1068  After so many requests, this is Bretagne. She ...                  11
1120  Say hello to this unbelievably well behaved sq...                 170
1165  Happy 4/20 from the squad! 13/10 for all https...                  20
1202  This is Bluebert. He just saw that both #Final...                  50
1228  Happy Saturday here's 9 puppers on a bench. 99...                  90
1254  Here's a brigade of puppers. All look very pre...                  80
1274  From left to right:\nCletus, Jerome, Alejandro...                  50
1351  Here is a whole flock of puppers.  60/50 I'll ...                  50
1433  Happy Wednesday here's a bucket of pups. 44/40...                  40
1598  Yes I do realize a rating of 4/20 would've bee...                  20
1634  Two sneaky puppers were not initially seen, mo...                 130
1635  Someone help the girl is being mugged. Several...                 110
1662  This is Darrel. He just robbed a 7/11 and is i...                  11
1663  I'm aware that I could've said 20/16, but here...                  16
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                 120
1843  Here we have an entire platoon of puppers. Tot...                  80
2335  This is an Albanian 3 1/2 legged  Episcopalian...                   2

In [17]:
data.tweet_id.duplicated().sum()

0

In [18]:
data.name.value_counts()

None            745
a                55
Charlie          12
Oliver           11
Cooper           11
Lucy             11
Tucker           10
Penny            10
Lola             10
Winston           9
Bo                9
the               8
Sadie             8
an                7
Buddy             7
Bailey            7
Toby              7
Daisy             7
Bella             6
Scout             6
Rusty             6
Dave              6
Jax               6
Jack              6
Stanley           6
Oscar             6
Leo               6
Milo              6
Koda              6
Larry             5
               ... 
incredibly        1
Christoper        1
Adele             1
Kona              1
Luther            1
Autumn            1
Alexanderson      1
light             1
Zuzu              1
Skittle           1
Ralphé            1
Mo                1
Nigel             1
Miguel            1
Tilly             1
Grizz             1
Kawhi             1
Major             1
Gerbald           1


In [19]:
image_pred.sample(5)

tweet_id                                          jpg_url  \
221   670093938074779648  https://pbs.twimg.com/media/CUym4Y5WsAEiI9_.jpg   
1417  771171053431250945  https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg   
840   694669722378485760  https://pbs.twimg.com/media/CaP2bS8WYAAsMdx.jpg   
778   689877686181715968  https://pbs.twimg.com/media/CZLwGAIWQAIYsTx.jpg   
515   676263575653122048  https://pbs.twimg.com/media/CWKSIfUUYAAiOBO.jpg   

      img_num                    p1   p1_conf  p1_dog                p2  \
221         1            toy_poodle  0.383346    True  miniature_poodle   
1417        3               Samoyed  0.978833    True        Pomeranian   
840         2                beaver  0.457094   False          mongoose   
778         1  Old_English_sheepdog  0.269155    True   Tibetan_terrier   
515         1                 teddy  0.098283   False        toy_poodle   

       p2_conf  p2_dog                p3   p3_conf  p3_dog  
221   0.153678    True              chow  0.138543    True  
1417  0.012763    True        Eskimo_dog  0.001853    True  
840   0.228298   False            marmot  0.148309   False  
778   0.111496    True  Lakeland_terrier  0.104939    True  
515   0.098029    True   shopping_basket  0.077852   False

In [20]:
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [21]:
image_pred.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [22]:
image_pred.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
cocker_spaniel                30
malamute                      30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
German_shepherd               20
Siberian_husky                20
Staffordshire_bullterrier     20
web_site                      19
Cardigan                      19
teddy                         18
Maltese_dog                   18
beagle                        18
Shetland_sheepdog             18
Eskimo_dog                    18
Lakeland_terrier              17
Shih-Tzu                      17
Rottweiler                    17
Italian_greyhound             16
kuvasz                        16
          

In [23]:
image_pred.tweet_id.duplicated().sum()

0

In [24]:
extra.tail()

contributors  coordinates          created_at display_text_range  \
2326           NaN          NaN 2015-11-16 00:24:50           [0, 120]   
2327           NaN          NaN 2015-11-16 00:04:52           [0, 137]   
2328           NaN          NaN 2015-11-15 23:21:54           [0, 130]   
2329           NaN          NaN 2015-11-15 23:05:30           [0, 139]   
2330           NaN          NaN 2015-11-15 22:32:08           [0, 131]   

                                               entities  \
2326  {'hashtags': [], 'symbols': [], 'user_mentions...   
2327  {'hashtags': [], 'symbols': [], 'user_mentions...   
2328  {'hashtags': [], 'symbols': [], 'user_mentions...   
2329  {'hashtags': [], 'symbols': [], 'user_mentions...   
2330  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  favorite_count  \
2326  {'media': [{'id': 666049244999131136, 'id_str'...              99   
2327  {'media': [{'id': 666044217047650304, 'id_str'...             274   
2328  {'media': [{'id': 666033409081393153, 'id_str'...             114   
2329  {'media': [{'id': 666029276303482880, 'id_str'...             121   
2330  {'media': [{'id': 666020881337073664, 'id_str'...            2446   

      favorited                                          full_text  geo  \
2326      False  Here we have a 1949 1st generation vulpix. Enj...  NaN   
2327      False  This is a purebred Piers Morgan. Loves to Netf...  NaN   
2328      False  Here is a very happy pup. Big fan of well-main...  NaN   
2329      False  This is a western brown Mitsubishi terrier. Up...  NaN   
2330      False  Here we have a Japanese Irish Setter. Lost eye...  NaN   

                            ...                          quoted_status  \
2326                        ...                                    NaN   
2327                        ...                                    NaN   
2328                        ...                                    NaN   
2329                        ...                                    NaN   
2330                        ...                                    NaN   

      quoted_status_id quoted_status_id_str  quoted_status_permalink  \
2326               NaN                  NaN                      NaN   
2327               NaN                  NaN                      NaN   
2328               NaN                  NaN                      NaN   
2329               NaN                  NaN                      NaN   
2330               NaN                  NaN                      NaN   

      retweet_count  retweeted  retweeted_status  \
2326             41      False               NaN   
2327            133      False               NaN   
2328             41      False               NaN   
2329             43      False               NaN   
2330            469      False               NaN   

                                                 source truncated  \
2326  <a href="http://twitter.com/download/iphone" r...     False   
2327  <a href="http://twitter.com/download/iphone" r...     False   
2328  <a href="http://twitter.com/download/iphone" r...     False   
2329  <a href="http://twitter.com/download/iphone" r...     False   
2330  <a href="http://twitter.com/download/iphone" r...     False   

                                                   user  
2326  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2327  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2328  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2329  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2330  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

In [25]:
extra.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 32 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2331 non-null datetime64[ns]
display_text_range               2331 non-null object
entities                         2331 non-null object
extended_entities                2059 non-null object
favorite_count                   2331 non-null int64
favorited                        2331 non-null bool
full_text                        2331 non-null object
geo                              0 non-null float64
id                               2331 non-null int64
id_str                           2331 non-null int64
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null float64
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 n

In [26]:
extra.dtypes

contributors                            float64
coordinates                             float64
created_at                       datetime64[ns]
display_text_range                       object
entities                                 object
extended_entities                        object
favorite_count                            int64
favorited                                  bool
full_text                                object
geo                                     float64
id                                        int64
id_str                                    int64
in_reply_to_screen_name                  object
in_reply_to_status_id                   float64
in_reply_to_status_id_str               float64
in_reply_to_user_id                     float64
in_reply_to_user_id_str                 float64
is_quote_status                            bool
lang                                     object
place                                    object
possibly_sensitive                      

In [27]:
extra.id.duplicated().sum()

0

### Quality issues

**Data table**
- 181 no original tweets (retweets) in table
- tweet_id should be a string, not an int. The same for other id columns.
- timestamp should be a datetime and not an object. The same for retweeted_status_timestamp.
- innacurate rating_denominator values for 23 tweets.
- rating_numerator has some really big numbers but there are no limit for the rating, needs further analysis.
- a lot of entries doesn't have a dog stage assigned. None used insteed of NaN.
- wrong name assigments: a, an, the. None used insteed of NaN.

**Image_pred table**
- no all the tweets in the data table have a image prediction.
- tweet_id should be a string, not an int.
- predictions that are not dogs: seat belt, web site, flamingo, minibus, fountain, limousine, among others.
- inconsistency on the cases (lowercase, uppercase) in the predictions.

**Extra table**
- tweet_id should be a string, not an int.
- no all the tweets in the data table is in this table.

### Tidiness issues
- The doggo, floofer, pupper, puppo columns should be only one column.
- There should be two tables: one for the tweet values and other for the dog features.
- Since the extra table has all the data from the tweet there are a lot of repeated information.